### PR TITLE
Add merlin support that doesn't require dune/jbuild

### LINKED ISSUE #273: Usable inside bucklescript repository
I (and other people I've talked to) have found it really difficult to work on the BuckleScript project directly because the language server doesn't detect the BuckleScript repository as a valid project.  Is there something we could do to special-case that repository so more people can contribute on it?

### PR DESCRIPTION
**Pull request by phated**
This allows RLS to support working on the BuckleScript project directly. I think there's a lot of improvements to be made but this will allow us to use the latest RLS to dev against the bucklescript repo.

Fixes #273 

cc @bobzhang @cristianoc

### CODE REVIEW

**Comment by bobzhang** [comment]
From the failure we got, I suggest that RLS be more forgiving, missing cmt file should not result in a fatal error

**Comment by jaredly** [comment]
Cool, thanks for putting this together! I wonder if `Merlin` is a misnomer in this case, given that `getCompiler` expects you to be very specifically working in the `bucklescript` project.

**Comment by jaredly** [comment]
@bobzhang could you explain what you mean?

File: src/analyze/BuildSystem.re [hunk 1]
```diff
@@ -302,6 +308,10 @@ let getCompiler = (rootPath, buildSystem) => {
     | Dune(Opam(switchPrefix)) =>
       let%try_wrap binPath = getOpamLibOrBinPath(rootPath, switchPrefix, "bin" /+ "ocamlopt.opt");
       binPath
+    | Merlin(rootPath) =>
+      // Assume OCaml is installed next to the project we are building
+      // This is due to BuckleScript shipping an OCaml compiler next to jscomp/
+      RResult.Ok(rootPath /+ ".." /+ "ocaml" /+ "ocamlc.opt")
```
Comment (jaredly) [inline src/analyze/BuildSystem.re]: Is there any way to not hardcode this? Can it be determined from the `.merlin` file, ideally?

Comment (phated) [reply src/analyze/BuildSystem.re]: Can you explain any further? I don't really know anything about Merlin

Comment (phated) [reply src/analyze/BuildSystem.re]: @bobzhang @jaredly @cristianoc do any of you understand what I'd need to do for this? I need some guidance.

**Comment by phated** [comment]
> bobzhang could you explain what you mean?

@jaredly Hongbo and I paired at ReasonConf to determine that these changes were failing inside the BuckleScript repo because there were 2-3 old files that weren't being built and RLS was crashing because it couldn't find artifacts for those files. I believe he is suggesting that RLS should be more fault tolerant in those cases, but I don't have the knowledge to fix/implement that.

File: src/analyze/BuildSystem.re (hunk 1 above)
Comment (bobzhang) [reply src/analyze/BuildSystem.re]: Can the path of ocaml compiler be configured so that it does not have to be hard coded here?

**Comment by phated** [comment]
How can I move this forward?

**Comment by phated** [comment]
@jaredly friendly ping. I'd love some concrete advice on what to do for this to land.

**Comment by jaredly** [comment]
Hi @phated, thanks for your persistence :) I'd be happy to work with you on getting rls working for the bucklescript project.
What I'd want in order for this to land:
- rename `Merlin` to `BucklescriptSource` or something, to indicate that we only expect this to be one specific project
- change the detection for a `BucklescriptSource` build type to be "has a .merlin, and has the ocaml binary in the expected adjacent spot"

How does that sound?

**Comment by phated** [comment]
I've just been using https://github.com/ocaml/ocaml-lsp and https://github.com/ocamllabs/vscode-ocaml-platform for my ocaml work, so closing this.

**Pull request closed without merging**